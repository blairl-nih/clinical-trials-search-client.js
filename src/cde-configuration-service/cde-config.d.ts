/**
 * Declaration for the CDEConfig module.  The actual CDEConfig module is an externally defined UMD module.
 * 
 * Example definition:
 * 
 *   // Uses AMD or browser globals to create a module.
 *   
 *   // Based on https://github.com/umdjs/umd/blob/master/templates/amdWeb.js
 *   
 *   // Defines a module "cdeConfig" that has no dependencies.
 *   // Note that the name of the module is implied by the file name. It is best
 *   // if the file name and the exported global have matching names.
 *   
 *   // If you do not want to support the browser global path, then you
 *   // can remove the `root` use and the passing of `this` as the first arg to
 *   // the top function.
 *   
 *   (function (root, factory) {
 *       if (typeof define === 'function' && define.amd) {
 *           // AMD. Register as an anonymous module.
 *           define(factory);
 *       } else {
 *           // Browser globals
 *           root.cdeConfig = factory();
 *       }
 *   }(this, function () {
 *       // Just return a value to define the module export.
 *       // This example returns an object, but the module
 *       // can return a function as the exported value.
 *       return {
 *           'environmentConfig': {
 *               'www.cancer.gov': {
 *                   configItem1: 'value1',
 *                   configItem2: 'value2'
 *               },
 *               'www-stage.cancer.gov': {
 *                   configItem2: 'value2-a'
 *               }
 *           }
 *       };
 *   }))
 *   
 */
declare module "CDEConfig" {
    export = CDEConfig;
}

declare const CDEConfig : any;
