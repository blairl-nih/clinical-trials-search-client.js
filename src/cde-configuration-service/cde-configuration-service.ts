import { CDEConfiguration } from '.';

/**
 * Speicifies an interface for retrieving environment-specific JavaScript configuration values.
 * This allows configuration settings to be specified at runtime rather than being baked into
 * the JavaScript at build-time.
 */
export interface CDEConfigurationService {

    /**
     * Gets the runtime configuration options for the current runtime environment.
     * (eg. A service URL specific to the lower tiers.)
     */
    getConfiguration() : CDEConfiguration;
}