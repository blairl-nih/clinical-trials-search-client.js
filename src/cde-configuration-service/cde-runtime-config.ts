import { CDEConfigurationService, CDEConfiguration, CDEDefaultConfiguration } from '.';
export class CDERuntimeConfig implements CDEConfigurationService {

    private defaultConfiguration : CDEConfiguration;

    constructor() {
        this.defaultConfiguration = new CDEDefaultConfiguration();
    }

    /**
     * Gets the runtime configuration options for the current runtime environment.
     * (eg. A service URL specific to the lower tiers.)
     */
    getConfiguration() : CDEConfiguration {

        // Get the default configuration.  Maintained in the CDEDefaultConfiguration model.
        let configuration:CDEConfiguration = this.defaultConfiguration;

        // If any overrides are available for the current runtime environment, load them on top of
        // the existing defaults.
        if (typeof CDEConfig !== 'undefined') {

            // Find environment-specific overrides.
            let hostname:string = window.location.hostname;
            if( hostname in CDEConfig.environmentConfig ) {
                console.log( 'Loading overrides for \'' + hostname + '\'');
                let overrides:CDEConfiguration = <CDEConfiguration>CDEConfig.environmentConfig[hostname];
                configuration = { ...this.defaultConfiguration, ...overrides };
            }

        }

        return configuration;
    }
}