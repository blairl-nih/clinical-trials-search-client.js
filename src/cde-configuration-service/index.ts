export * from './model';
export * from './cde-configuration-service';
export * from './cde-runtime-config';
