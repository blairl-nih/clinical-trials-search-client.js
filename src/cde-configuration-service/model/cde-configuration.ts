/**
 * Interface for specifying properties to store configuration options.
 * 
 * NOTE: New options must be defined in CDEConfiguration and given explicit default values
 * in CDEDefaultConfiguration.
 */
export interface CDEConfiguration {
    /**
     * Identifies the hostname of the clinicaltrials API server
     * 
     * @type {string}
     * @memberof CDEConfiguration
     */
    ClinicalTrialsAPIServer : string;

    /**
     * The key required to conect to the Google developer API
     * 
     * @type {string}
     * @memberof CDEConfiguration
     */
    GoogleAPIKey : string;    

    /**
     * Identifies the hostname of the R4R API server
     * 
     * @type {string}
     * @memberof CDEConfiguration
     */
    R4RAPIServer : string;    

}