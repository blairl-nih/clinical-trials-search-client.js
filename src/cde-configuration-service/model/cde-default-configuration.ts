import { CDEConfiguration } from '.';

/**
 * Class for specifying default values for runtime configuration options.
 * 
 * NOTE: New options must be defined in CDEConfiguration and given explicit default values
 * in CDEDefaultConfiguration.
 */
export class CDEDefaultConfiguration implements CDEConfiguration {
    /**
     * Identifies the hostname of the clinicaltrials API server
     * 
     * @type {string}
     * @memberof CDEDefaultConfiguration
     */
    ClinicalTrialsAPIServer : string = 'clinicaltrialsapi.cancer.gov';

    /**
     * The key required to conect to the Google developer API
     * 
     * @type {string}
     * @memberof CDEDefaultConfiguration
     */
    // This is our official key
    GoogleAPIKey : string = 'AIzaSyB0DtmqLHmCLjAMjGfNh9myktZYz2-yIfg';
    
    // This is Brendan's key for testing on localhost
    // GoogleAPIKey : string = 'AIzaSyASmav6-lTS2G7xb-7fL146lsvtK9_oJgs';
 
        /**
     * Identifies the hostname of the clinicaltrials API server
     * 
     * @type {string}
     * @memberof CDEDefaultConfiguration
     */
    R4RAPIServer : string = 'https://webapis.cancer.gov/r4r/v1';
}
