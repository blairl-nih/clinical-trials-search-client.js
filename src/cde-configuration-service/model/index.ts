export * from './cde-configuration';
export * from './cde-default-configuration';