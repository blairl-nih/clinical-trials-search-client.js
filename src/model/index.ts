export * from './term-result';
export * from './term-results';
export * from './intervention-result';
export * from './intervention-results';
export * from './disease-result';
export * from './disease-results';