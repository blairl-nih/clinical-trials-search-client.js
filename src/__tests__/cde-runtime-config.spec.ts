import { CDERuntimeConfig } from "../cde-configuration-service/cde-runtime-config";
import { CDEConfiguration } from "../cde-configuration-service/model/cde-configuration";

describe("Services.CDEConfigurationService.CDERuntimeConfig", () => {
  it("should return defaults when no config loaded", () => {
    let config: CDERuntimeConfig = new CDERuntimeConfig();
    let actual: CDEConfiguration = config.getConfiguration();
    let expected: CDEConfiguration = <CDEConfiguration>{
      ClinicalTrialsAPIServer: "clinicaltrialsapi.cancer.gov",
      GoogleAPIKey: "AIzaSyB0DtmqLHmCLjAMjGfNh9myktZYz2-yIfg",
      R4RAPIServer: "https://webapis.cancer.gov/r4r/v1"
    };

    expect(actual).toMatchSnapshot();
  });

  it("should override defaults", () => {
    // We need 2 things to make this work:
    // 1. A CDE config object - need to figure out how to add object to global
    // 2. A fake hostname - use sinon or pushstate?
    // See: http://www.bradoncode.com/blog/2015/02/27/karma-tutorial/

    // Create new config object
    let config: CDERuntimeConfig = new CDERuntimeConfig();

    // Get the configuration (either default or overrides if they exist)
    let actual: CDEConfiguration = config.getConfiguration();

    // Set value to compare against
    let expected: CDEConfiguration = <CDEConfiguration>{
      ClinicalTrialsAPIServer: "clinicaltrialsapi.cancer.gov",
      GoogleAPIKey: "AIzaSyB0DtmqLHmCLjAMjGfNh9myktZYz2-yIfg",
      R4RAPIServer: "https://webapis.cancer.gov/r4r/v1"
    };

    expect(actual).toMatchSnapshot();
  });
});
