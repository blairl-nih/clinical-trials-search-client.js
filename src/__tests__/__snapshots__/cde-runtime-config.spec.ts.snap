// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services.CDEConfigurationService.CDERuntimeConfig should override defaults 1`] = `
CDEDefaultConfiguration {
  "ClinicalTrialsAPIServer": "clinicaltrialsapi.cancer.gov",
  "GoogleAPIKey": "AIzaSyB0DtmqLHmCLjAMjGfNh9myktZYz2-yIfg",
  "R4RAPIServer": "https://webapis.cancer.gov/r4r/v1",
}
`;

exports[`Services.CDEConfigurationService.CDERuntimeConfig should return defaults when no config loaded 1`] = `
CDEDefaultConfiguration {
  "ClinicalTrialsAPIServer": "clinicaltrialsapi.cancer.gov",
  "GoogleAPIKey": "AIzaSyB0DtmqLHmCLjAMjGfNh9myktZYz2-yIfg",
  "R4RAPIServer": "https://webapis.cancer.gov/r4r/v1",
}
`;
